src/lint/rules/variable-use-def.ts

```diff
@@ -0,0 +1,352 @@
+      preceding.textContent != null
+    ) {
+      // `__` is for <del>_x_</del><ins>_y_</ins>, which has textContent `_x__y_`
+      for (const name of preceding.textContent.matchAll(/(?<=\b|_)_([a-zA-Z0-9]+)_(?=\b|_)/g)) {
+        scope.declare(name[1], null);
+      }
+    }
+    preceding = previousOrParent(preceding, parentClause);
+  }
+
+  walkAlgorithm(steps, scope, report);
+
+  for (const [name, { kind, used, node }] of scope.vars) {
+    if (!used && node != null && kind !== 'parameter' && kind !== 'abstract closure parameter') {
+      // prettier-ignore
+      const message = `${JSON.stringify(name)} is declared here, but never referred to`;
+      report({
+        ruleId: 'unused-declaration',
+        message,
+        line: node.location.start.line,
+        column: node.location.start.column,
+      });
+    }
+  }
+}
+
+function walkAlgorithm(steps: OrderedListNode | UnorderedListNode, scope: Scope, report: Reporter) {
+  for (const step of steps.contents) {
+    const parts = step.contents;
+    const loopVars: Set<VariableNode> = new Set();
+    const declaredThisLine: Set<VariableNode> = new Set();
+
+    let firstRealIndex = 0;
+    let first = parts[firstRealIndex];
+    while (['comment', 'tag', 'opaqueTag'].includes(first.name)) {
+      ++firstRealIndex;
+      first = parts[firstRealIndex];
+    }
+
+    let lastRealIndex = parts.length - 1;
+    let last = parts[lastRealIndex];
+    while (['comment', 'tag', 'opaqueTag'].includes(last.name)) {
+      --lastRealIndex;
+      last = parts[lastRealIndex];
+    }
+
+    // handle [declared="foo"] attributes
+    const extraDeclarations = step.attrs.find(d => d.key === 'declared');
+    if (extraDeclarations != null) {
+      for (let name of extraDeclarations.value.split(',')) {
+        name = name.trim();
+        const line = extraDeclarations.location.start.line;
+        const column =
+          extraDeclarations.location.start.column +
+          extraDeclarations.key.length +
+          2 + // '="'
+          findDeclaredAttrOffset(extraDeclarations.value, name);
+        if (scope.declared(name)) {
+          // prettier-ignore
+          const message = `${JSON.stringify(name)} is already declared and does not need an explict annotation`;
+          report({
+            ruleId: 'unnecessary-declared-var',
+            message,
+            line,
+            column,
+          });
+        } else {
+          scope.declare(name, { location: { start: { line, column } } }, 'attribute declaration');
+        }
+      }
+    }
+
+    // handle loops
+    if (first.name === 'text' && first.contents.startsWith('For each ')) {
+      let loopVar = parts[firstRealIndex + 1];
+      if (loopVar?.name === 'pipe') {
+        loopVar = parts[firstRealIndex + 3]; // 2 is the space
+      }
+      if (isVariable(loopVar)) {
+        loopVars.add(loopVar);
+
+        scope.declare(loopVar.contents[0].contents, loopVar, 'loop variable');
```

Comment:
`break` here so you don't try to declare the loop target or anything mentioned in the guard

Reply:
We don't want to _declare_ anything else, but we do want to count references on the rest of the line as uses, and just putting a `continue` here would skip that. What test are you imagining `continue`ing here would fix?

Reply:
```
1. For each integer _a_ of _b_ such that _c_ relates to _a_ in some way, do
```

Reply:
That case is guarded explicitly: https://github.com/tc39/ecmarkup/blob/23e5a62eb15efdb3cb726577ed19d54d19f91c25/src/lint/rules/variable-use-def.ts#L294.

Reply:
There's a test for it too: https://github.com/tc39/ecmarkup/blob/23e5a62eb15efdb3cb726577ed19d54d19f91c25/test/lint-variable-use-def.js#L102-L116

Though not using the "for each" style; I can add one for that case, if you want.

Reply:
Yes, please.

Reply:
Done.